pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/properties/PreserveStashesJobProperty.java

```diff
@@ -0,0 +1,167 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class PreserveStashesJobProperty extends OptionalJobProperty<WorkflowJob> {
+
+    private static final Logger LOGGER = Logger.getLogger(PreserveStashesJobProperty.class.getName());
+
+    private static final int MAX_SAVED_STASHES = 50;
+
+    private int buildCount = 1;
+
+    @DataBoundConstructor
+    public PreserveStashesJobProperty() {
+
+    }
+
+    public int getBuildCount() {
+        return buildCount;
+    }
+
+    @DataBoundSetter
+    public void setBuildCount(int buildCount) {
+        if (buildCount > MAX_SAVED_STASHES) {
+            throw new IllegalArgumentException("buildCount must be between 1 and " + MAX_SAVED_STASHES);
+        }
+        this.buildCount = buildCount;
+    }
+
+    @Extension @Symbol("preserveStashes")
+    public static class DescriptorImpl extends OptionalJobPropertyDescriptor {
+        @Override
+        @Nonnull
+        public String getDisplayName() {
+            return "Preserve stashes from completed builds";
+        }
+
+        public FormValidation doCheckBuildCount(@QueryParameter int value) {
+            if (value < 0) {
+                return FormValidation.error("Must be greater than or equal to 0");
+            } else if (value > 50) {
+                return FormValidation.error("Must be 50 or less");
+            } else {
+                return FormValidation.ok();
+            }
+        }
+    }
+
+    @Extension
+    public static final class SaveStashes extends StashManager.StashBehavior {
+
+        @Override
+        public boolean shouldClearAll(@Nonnull Run<?,?> build) {
+            if (build instanceof WorkflowRun) {
+                WorkflowRun r = (WorkflowRun)build;
+                WorkflowJob j = r.getParent();
+                PreserveStashesJobProperty prop = j.getProperty(PreserveStashesJobProperty.class);
+                if (prop != null) {
+                    int bc = prop.getBuildCount();
+                    if (bc > 0) {
+                        for (WorkflowRun recentRun : j.getBuilds().completedOnly().limit(bc)) {
+                            if (recentRun != null && recentRun.equals(r)) {
+                                return false;
+                            }
+                        }
+                    }
+                }
+            }
+            return true;
+        }
+    }
+
+    @Extension
+    public static final class StashClearingListener extends RunListener<Run<?,?>> {
+        @Override
+        public void onCompleted(Run<?,?> r, TaskListener listener) {
+            if (r instanceof WorkflowRun) {
+                WorkflowJob j = ((WorkflowRun) r).getParent();
+                PreserveStashesJobProperty prop = j.getProperty(PreserveStashesJobProperty.class);
+                if (prop != null && prop.getBuildCount() > 0) {
+                    // The "+1" is to ensure that we look at one more than the maximum possible number of builds with saved stashes.
```

Comment:
Hmm...maybe? They'd all have to complete at basically the exact same time for that to even theoretically happen. Yeah, I don't think it can plausibly happen.